src/core/reader/types/element.rs

```diff
@@ -0,0 +1,164 @@
+use core::fmt::Debug;
+
+use alloc::vec::Vec;
+
+use crate::{
+    core::reader::span::Span, read_constant_expression::read_constant_instructions, Error, Result,
+};
+
+use super::RefType;
+
+#[derive(Clone)]
+pub struct ElemType {
+    pub init: ElemItems,
+    pub mode: ElemMode,
+}
```

Comment:
I am strongly in favor of this to be in this form for better correspondence with the spec and understandability:
```
pub struct ElemSegment {
  pub ty: RefType,
  pub init: Vec<Span>,
  pub mode: ElemMode
}
```
This can be created during element section read of validation::validate() and also validated.
and then the bunch of ElemSegments could be instantiated to ElemInst's within Store::init_store() (with the const spans evaluated), almost exactly like Globals to GlobalInsts. I think treating them like a mixture of DataSegments and Globals within our implementation feels like a such a perfect fit, I would feel bad glossing over it.

Reply:
I'm not 100% convinced that this issue is critical enough for the effort needed to refactor this. Also, having to convert u32 (in the case of reffuncs) into initialization expressions seems inefficient.
- For `ty = RefFuncs`, we  can have the indicies directly in the binary `u32` or as expressions, which isn't really represented your proposal.
- For `ty = ExternFuncs` we only have expressions (afaik).

Reply:
fair enough :)